File: init-service/app/services/node_service.py
import os
import shutil
import uuid
from datetime import UTC, datetime

from fastapi import HTTPException, status
from sqlalchemy import select

from app.db import session_scope
from app.models.project import Node
from app.services.cas_service import hash_content, hash_file, upload_file_if_missing, upload_if_missing
from app.services.workspace_service import (
    get_node,
    project_disk_path,
    relative_path_for_node,
    serialize_node,
)

LIVE_SYNC_IGNORE_NAMES = {".git", "__pycache__", ".pytest_cache"}


def _now() -> datetime:
    return datetime.now(UTC)


def _validate_name(name: str) -> str:
    normalized = name.strip()
    if not normalized:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Name cannot be empty.")
    if "/" in normalized or "\\" in normalized:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Name cannot contain path separators.")
    return normalized


def _node_abs_path(session, node: Node) -> str:
    rel_path = relative_path_for_node(session, node)
    return os.path.join(project_disk_path(node.project.workspace_id, node.project_id), rel_path) if rel_path else project_disk_path(node.project.workspace_id, node.project_id)


def _assert_folder(node: Node) -> None:
    if node.type != "FOLDER":
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Target node must be a folder.")


def _assert_file(node: Node) -> None:
    if node.type != "FILE":
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Target node must be a file.")


def _content_hash_for_path(abs_path: str, existing_node: Node | None = None) -> tuple[str | None, int | None]:
    try:
        size_bytes = os.path.getsize(abs_path)
    except OSError:
        return None, None

    if existing_node and existing_node.content_hash and existing_node.size_bytes == size_bytes:
        return existing_node.content_hash, existing_node.size_bytes

    try:
        content_hash, size_bytes = hash_file(abs_path)
    except OSError:
        return None, None

    upload_file_if_missing(content_hash, abs_path)
    return content_hash, size_bytes


def _sync_folder_children(session, parent: Node) -> None:
    abs_path = _node_abs_path(session, parent)
    if not os.path.isdir(abs_path):
        return

    existing_children = {
        child.name: child
        for child in session.scalars(
            select(Node).where(Node.parent_id == parent.id)
        ).all()
    }

    discovered_entries: list[os.DirEntry[str]] = []
    with os.scandir(abs_path) as entries:
        for entry in entries:
            if entry.name in LIVE_SYNC_IGNORE_NAMES:
                continue
            discovered_entries.append(entry)

    discovered_names = {entry.name for entry in discovered_entries}
    for child_name, child in existing_children.items():
        if child_name not in discovered_names:
            session.delete(child)
    session.flush()

    now = _now()
    for entry in sorted(discovered_entries, key=lambda current: (not current.is_dir(follow_symlinks=False), current.name.lower())):
        existing = existing_children.get(entry.name)

        if entry.is_dir(follow_symlinks=False):
            if existing is None:
                session.add(
                    Node(
                        id=str(uuid.uuid4()),
                        project_id=parent.project_id,
                        parent_id=parent.id,
                        name=entry.name,
                        type="FOLDER",
                        content_hash=None,
                        size_bytes=None,
                        created_at=now,
                        updated_at=now,
                    )
                )
                continue

            if existing.type != "FOLDER":
                existing.type = "FOLDER"
                existing.content_hash = None
                existing.size_bytes = None
                existing.updated_at = now
            continue

        if not entry.is_file(follow_symlinks=False):
            continue

        content_hash, size_bytes = _content_hash_for_path(entry.path, existing if existing and existing.type == "FILE" else None)
        if existing is None:
            session.add(
                Node(
                    id=str(uuid.uuid4()),
                    project_id=parent.project_id,
                    parent_id=parent.id,
                    name=entry.name,
                    type="FILE",
                    content_hash=content_hash,
                    size_bytes=size_bytes,
                    created_at=now,
                    updated_at=now,
                )
            )
            continue

        if existing.type != "FILE":
            existing.type = "FILE"
            existing.updated_at = now

        if existing.content_hash != content_hash or existing.size_bytes != size_bytes:
            existing.content_hash = content_hash
            existing.size_bytes = size_bytes
            existing.updated_at = now

    session.flush()


def get_children(owner_user_id: str, node_id: str) -> dict:
    with session_scope() as session:
        node = get_node(session, node_id, owner_user_id)
        _assert_folder(node)
        _sync_folder_children(session, node)
        children = session.scalars(
            select(Node)
            .where(Node.parent_id == node.id)
            .order_by(Node.type.desc(), Node.name)
        ).all()
        return {"nodes": [serialize_node(session, child) for child in children]}


def read_content(owner_user_id: str, node_id: str) -> dict:
    with session_scope() as session:
        node = get_node(session, node_id, owner_user_id)
        _assert_file(node)
        abs_path = _node_abs_path(session, node)
        try:
            with open(abs_path, "r", encoding="utf-8") as handle:
                content = handle.read()
        except FileNotFoundError:
            content = ""
        return {"node": serialize_node(session, node), "content": content}


def create_node(owner_user_id: str, parent_id: str, name: str, node_type: str) -> dict:
    with session_scope() as session:
        parent = get_node(session, parent_id, owner_user_id)
        _assert_folder(parent)
        node_name = _validate_name(name)
        now = _now()
        node = Node(
            id=str(uuid.uuid4()),
            project_id=parent.project_id,
            parent_id=parent.id,
            name=node_name,
            type=node_type,
            content_hash=None,
            size_bytes=None,
            created_at=now,
            updated_at=now,
        )
        session.add(node)
        session.flush()

        abs_path = _node_abs_path(session, node)
        if node_type == "FOLDER":
            os.makedirs(abs_path, exist_ok=True)
        else:
            os.makedirs(os.path.dirname(abs_path), exist_ok=True)
            with open(abs_path, "w", encoding="utf-8") as handle:
                handle.write("")
        session.flush()
        return serialize_node(session, node)


def update_content(owner_user_id: str, node_id: str, content: str) -> dict:
    payload = content.encode("utf-8")
    content_hash = hash_content(payload)
    upload_if_missing(content_hash, payload)

    with session_scope() as session:
        node = get_node(session, node_id, owner_user_id)
        _assert_file(node)
        abs_path = _node_abs_path(session, node)
        os.makedirs(os.path.dirname(abs_path), exist_ok=True)
        with open(abs_path, "w", encoding="utf-8") as handle:
            handle.write(content)
        node.content_hash = content_hash
        node.size_bytes = len(payload)
        node.updated_at = _now()
        session.flush()
        return serialize_node(session, node)


def rename_node(owner_user_id: str, node_id: str, name: str) -> dict:
    with session_scope() as session:
        node = get_node(session, node_id, owner_user_id)
        if node.parent_id is None:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Project root cannot be renamed.")
        old_abs_path = _node_abs_path(session, node)
        node.name = _validate_name(name)
        node.updated_at = _now()
        session.flush()
        new_abs_path = _node_abs_path(session, node)
        os.makedirs(os.path.dirname(new_abs_path), exist_ok=True)
        if os.path.exists(old_abs_path):
            os.replace(old_abs_path, new_abs_path)
        return serialize_node(session, node)


def move_node(owner_user_id: str, node_id: str, parent_id: str) -> dict:
    with session_scope() as session:
        node = get_node(session, node_id, owner_user_id)
        if node.parent_id is None:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Project root cannot be moved.")
        target_parent = get_node(session, parent_id, owner_user_id)
        _assert_folder(target_parent)
        if target_parent.project_id != node.project_id:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Cannot move nodes across projects.")

        current = target_parent
        while current.parent_id is not None:
            if current.id == node.id:
                raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Cannot move a folder into its own descendant.")
            current = get_node(session, current.parent_id, owner_user_id)

        old_abs_path = _node_abs_path(session, node)
        node.parent_id = target_parent.id
        node.updated_at = _now()
        session.flush()
        new_abs_path = _node_abs_path(session, node)
        os.makedirs(os.path.dirname(new_abs_path), exist_ok=True)
        if os.path.exists(old_abs_path):
            os.replace(old_abs_path, new_abs_path)
        return serialize_node(session, node)


def delete_node(owner_user_id: str, node_id: str) -> dict:
    with session_scope() as session:
        node = get_node(session, node_id, owner_user_id)
        if node.parent_id is None:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail="Project root cannot be deleted.")
        abs_path = _node_abs_path(session, node)
        serialized = serialize_node(session, node)
        session.delete(node)
        session.flush()
        if os.path.isdir(abs_path):
            shutil.rmtree(abs_path, ignore_errors=True)
        elif os.path.exists(abs_path):
            os.remove(abs_path)
        return serialized

File: init-service/src/api/node/node-service.ts
import fs from "fs";
import path from "path";
import { eq, and, desc } from "drizzle-orm";
import { db } from "../../loaders/postgres";
import { nodes as nodesTable } from "db-schema";
import ApiError from "../../utils/ApiError";
import { hashContent, hashFile, uploadFileIfMissing, uploadIfMissing } from "../../shared/cas-service";
import {
  getNode,
  projectDiskPath,
  relativePathForNode,
  serializeNode,
  validateName,
} from "../project/workspace-service";

const LIVE_SYNC_IGNORE_NAMES = new Set([".git", ".venv", "venv", "__pycache__", ".pytest_cache", "node_modules"]);

async function nodeAbsPath(tx: any, node: any): Promise<string> {
  const relPath = await relativePathForNode(node, tx);
  const projPath = projectDiskPath(node.project.workspace.id, node.projectId);
  return relPath ? path.join(projPath, relPath) : projPath;
}

function assertFolder(node: any): void {
  if (node.type !== "FOLDER") {
    throw new ApiError("Target node must be a folder.", 422);
  }
}

function assertFile(node: any): void {
  if (node.type !== "FILE") {
    throw new ApiError("Target node must be a file.", 422);
  }
}

async function contentHashForPath(absPath: string, existingNode?: any): Promise<[string | null, number | null]> {
  try {
    const stat = fs.statSync(absPath);
    const sizeBytes = stat.size;

    if (existingNode && existingNode.contentHash && existingNode.sizeBytes === sizeBytes) {
      return [existingNode.contentHash, existingNode.sizeBytes];
    }

    const [contentHash, calculatedSize] = await hashFile(absPath);
    await uploadFileIfMissing(contentHash, absPath);
    return [contentHash, calculatedSize];
  } catch {
    return [null, null];
  }
}

async function syncFolderChildren(tx: any, parent: any): Promise<void> {
  const absPath = await nodeAbsPath(tx, parent);
  if (!fs.existsSync(absPath) || !fs.statSync(absPath).isDirectory()) {
    return;
  }

  // Get existing database child nodes
  const dbChildren = await tx
    .select()
    .from(nodesTable)
    .where(eq(nodesTable.parentId, parent.id));

  const dbChildrenMap: Record<string, any> = {};
  for (const child of dbChildren) {
    dbChildrenMap[child.name] = child;
  }

  // Scan disk children
  const diskEntries = fs.readdirSync(absPath, { withFileTypes: true });
  const discoveredEntries = diskEntries.filter((entry) => !LIVE_SYNC_IGNORE_NAMES.has(entry.name));

  const discoveredNames = new Set(discoveredEntries.map((e) => e.name));

  // 1. Delete DB nodes not found on disk
  for (const childName of Object.keys(dbChildrenMap)) {
    if (!discoveredNames.has(childName)) {
      const child = dbChildrenMap[childName];
      await tx.delete(nodesTable).where(eq(nodesTable.id, child.id));
    }
  }

  const now = new Date();

  // 2. Add / update DB nodes matching disk
  // Sort folders first
  const sortedEntries = discoveredEntries.sort((a, b) => {
    if (a.isDirectory() !== b.isDirectory()) {
      return a.isDirectory() ? -1 : 1;
    }
    return a.name.toLowerCase().localeCompare(b.name.toLowerCase());
  });

  for (const entry of sortedEntries) {
    const existing = dbChildrenMap[entry.name];
    const entryPath = path.join(absPath, entry.name);

    if (entry.isDirectory()) {
      if (!existing) {
        await tx.insert(nodesTable).values({
          id: crypto.randomUUID(),
          projectId: parent.projectId,
          parentId: parent.id,
          name: entry.name,
          type: "FOLDER",
          contentHash: null,
          sizeBytes: null,
          createdAt: now,
          updatedAt: now,
        });
      } else if (existing.type !== "FOLDER") {
        await tx
          .update(nodesTable)
          .set({ type: "FOLDER", contentHash: null, sizeBytes: null, updatedAt: now })
          .where(eq(nodesTable.id, existing.id));
      }
      continue;
    }

    if (entry.isFile()) {
      const [contentHash, sizeBytes] = await contentHashForPath(
        entryPath,
        existing && existing.type === "FILE" ? existing : undefined
      );

      if (!existing) {
        await tx.insert(nodesTable).values({
          id: crypto.randomUUID(),
          projectId: parent.projectId,
          parentId: parent.id,
          name: entry.name,
          type: "FILE",
          contentHash,
          sizeBytes,
          createdAt: now,
          updatedAt: now,
        });
      } else {
        const needsUpdate =
          existing.type !== "FILE" ||
          existing.contentHash !== contentHash ||
          existing.sizeBytes !== sizeBytes;

        if (needsUpdate) {
          await tx
            .update(nodesTable)
            .set({ type: "FILE", contentHash, sizeBytes, updatedAt: now })
            .where(eq(nodesTable.id, existing.id));
        }
      }
    }
  }
}

// ─── Exported Services ────────────────────────────────────────────────────────

export async function getChildren(ownerUserId: string, nodeId: string): Promise<any> {
  return await db.transaction(async (tx) => {
    const node = await getNode(nodeId, ownerUserId, tx);
    assertFolder(node);
    await syncFolderChildren(tx, node);

    const children = await tx
      .select()
      .from(nodesTable)
      .where(eq(nodesTable.parentId, node.id))
      .orderBy(desc(nodesTable.type), nodesTable.name); // folders first, then alphabetical

    const serializedChildren = await Promise.all(children.map((c) => serializeNode(c, tx)));
    return { nodes: serializedChildren };
  });
}

export async function readContent(ownerUserId: string, nodeId: string): Promise<any> {
  return await db.transaction(async (tx) => {
    const node = await getNode(nodeId, ownerUserId, tx);
    assertFile(node);

    const absPath = await nodeAbsPath(tx, node);
    let content = "";

    try {
      // Check if file is text or binary
      const buf = fs.readFileSync(absPath);
      // Basic binary file detector (has null bytes)
      const isBinary = buf.includes(0);
      if (isBinary) {
        content = "BINARY_FILE";
      } else {
        content = buf.toString("utf-8");
      }
    } catch (e: any) {
      if (e.code !== "ENOENT") {
        throw e;
      }
      content = "";
    }

    const serializedNode = await serializeNode(node, tx);
    return { node: serializedNode, content };
  });
}

export async function createNode(ownerUserId: string, parentId: string, name: string, nodeType: string): Promise<any> {
  return await db.transaction(async (tx) => {
    const parent = await getNode(parentId, ownerUserId, tx);
    assertFolder(parent);

    const nodeName = validateName(name);
    const now = new Date();
    const nodeId = crypto.randomUUID();

    const node = {
      id: nodeId,
      projectId: parent.projectId,
      parentId: parent.id,
      name: nodeName,
      type: nodeType,
      contentHash: null,
      sizeBytes: null,
      createdAt: now,
      updatedAt: now,
    };

    await tx.insert(nodesTable).values(node);

    const absPath = await nodeAbsPath(tx, node);
    if (nodeType === "FOLDER") {
      fs.mkdirSync(absPath, { recursive: true });
    } else {
      fs.mkdirSync(path.dirname(absPath), { recursive: true });
      fs.writeFileSync(absPath, "");
    }

    return await serializeNode(node, tx);
  });
}

export async function updateContent(ownerUserId: string, nodeId: string, content: string): Promise<any> {
  const payload = Buffer.from(content, "utf-8");
  const contentHash = hashContent(payload);
  await uploadIfMissing(contentHash, payload);

  return await db.transaction(async (tx) => {
    const node = await getNode(nodeId, ownerUserId, tx);
    assertFile(node);

    const absPath = await nodeAbsPath(tx, node);
    fs.mkdirSync(path.dirname(absPath), { recursive: true });
    fs.writeFileSync(absPath, content);

    const now = new Date();
    await tx
      .update(nodesTable)
      .set({
        contentHash,
        sizeBytes: payload.length,
        updatedAt: now,
      })
      .where(eq(nodesTable.id, nodeId));

    const updatedNode = {
      ...node,
      contentHash,
      sizeBytes: payload.length,
      updatedAt: now,
    };

    return await serializeNode(updatedNode, tx);
  });
}

export async function renameNode(ownerUserId: string, nodeId: string, name: string): Promise<any> {
  return await db.transaction(async (tx) => {
    const node = await getNode(nodeId, ownerUserId, tx);
    if (!node.parentId) {
      throw new ApiError("Project root cannot be renamed.", 422);
    }

    const oldAbsPath = await nodeAbsPath(tx, node);
    const nodeName = validateName(name);
    const now = new Date();

    await tx
      .update(nodesTable)
      .set({ name: nodeName, updatedAt: now })
      .where(eq(nodesTable.id, nodeId));

    const updatedNode = { ...node, name: nodeName, updatedAt: now };
    const newAbsPath = await nodeAbsPath(tx, updatedNode);

    fs.mkdirSync(path.dirname(newAbsPath), { recursive: true });
    if (fs.existsSync(oldAbsPath)) {
      fs.renameSync(oldAbsPath, newAbsPath);
    }

    return await serializeNode(updatedNode, tx);
  });
}

export async function moveNode(ownerUserId: string, nodeId: string, parentId: string): Promise<any> {
  return await db.transaction(async (tx) => {
    const node = await getNode(nodeId, ownerUserId, tx);
    if (!node.parentId) {
      throw new ApiError("Project root cannot be moved.", 422);
    }

    const targetParent = await getNode(parentId, ownerUserId, tx);
    assertFolder(targetParent);

    if (targetParent.projectId !== node.projectId) {
      throw new ApiError("Cannot move nodes across projects.", 422);
    }

    // Check if moving to descendant
    let current = targetParent;
    while (current.parentId) {
      if (current.id === node.id) {
        throw new ApiError("Cannot move a folder into its own descendant.", 422);
      }
      current = await getNode(current.parentId, ownerUserId, tx);
    }

    const oldAbsPath = await nodeAbsPath(tx, node);
    const now = new Date();

    await tx
      .update(nodesTable)
      .set({ parentId: targetParent.id, updatedAt: now })
      .where(eq(nodesTable.id, nodeId));

    const updatedNode = { ...node, parentId: targetParent.id, updatedAt: now };
    const newAbsPath = await nodeAbsPath(tx, updatedNode);

    fs.mkdirSync(path.dirname(newAbsPath), { recursive: true });
    if (fs.existsSync(oldAbsPath)) {
      fs.renameSync(oldAbsPath, newAbsPath);
    }

    return await serializeNode(updatedNode, tx);
  });
}

export async function deleteNode(ownerUserId: string, nodeId: string): Promise<any> {
  return await db.transaction(async (tx) => {
    const node = await getNode(nodeId, ownerUserId, tx);
    if (!node.parentId) {
      throw new ApiError("Project root cannot be deleted.", 422);
    }

    const absPath = await nodeAbsPath(tx, node);
    const serialized = await serializeNode(node, tx);

    await tx.delete(nodesTable).where(eq(nodesTable.id, nodeId));

    if (fs.existsSync(absPath)) {
      fs.rmSync(absPath, { recursive: true, force: true });
    }

    return serialized;
  });
}
